Web Inspector: support console screenshots in a Worker

dcrousso opened the pull request:
#### de9f264928a4dceb1b3c975d2c8a63fa15f4f422
<pre>
Web Inspector: support console screenshots in a Worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=243361">https://bugs.webkit.org/show_bug.cgi?id=243361</a>
&lt;<a href="https://rdar.apple.com/problem/98223234">rdar://problem/98223234</a>&gt;

Reviewed by Darin Adler.

This will be very useful for developers that are working with `Blob` inside a `Worker` and don&apos;t have an easy way of visualizing it (e.g. sending/transferring to the webpage for rendering via the DOM).

* Source/WebCore/workers/WorkerConsoleClient.cpp:
(WebCore::canvasRenderingContext): Added.
(WebCore::WorkerConsoleClient::screenshot):
Support `ImageData`, `ImageBitmap`, `OffscreenCanvas`, all the different types of `CanvasRenderingContext`, and valid base64 `data:` URL strings.

* LayoutTests/inspector/console/console-screenshot.html:
* LayoutTests/inspector/worker/console-basic.html:
* LayoutTests/inspector/worker/console-basic-subworker.html:
* LayoutTests/inspector/worker/console-screenshot.html: Added.
* LayoutTests/inspector/worker/console-screenshot-expected.txt: Added.
* LayoutTests/inspector/worker/resources/worker-console.js:
(onmessage):

* Source/WebCore/page/FrameConsoleClient.cpp:
(WebCore::FrameConsoleClient::screenshot):

Drive-by: add some comments to clarify what each argument means.
Canonical link: <a href="https://commits.webkit.org/302778@main">https://commits.webkit.org/302778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acdc5d3a5f2648f620ad21e28715eece123cf5a8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130184 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41138 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137595 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81737 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132055 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2345 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99177 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133131 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116598 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79870 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34726 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80858 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110271 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35233 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140071 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2245 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2094 "19 flakes 1 failures") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107701 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2289 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112941 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107579 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27378 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1784 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31391 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55184 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2315 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65702 "Found 13 new failures in workers/WorkerConsoleClient.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2132 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2336 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2241 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/FrameConsoleClient.cpp:
```diff
@@ -468,11 +466,11 @@ void FrameConsoleClient::screenshot(JSC::JSGlobalObject* lexicalGlobalObject, Re
     }
 
     Vector<JSC::Strong<JSC::Unknown>> adjustedArguments;
+    adjustedArguments.reserveInitialCapacity(arguments->argumentCount());
```
This allocates initial capacity that is 1 too small when `target` is `nullptr`, resulting in a vector that is too big when the append overflows the initial capacity. Here is one at to correct that, but maybe you can find a better way.
```suggestion
    adjustedArguments.reserveInitialCapacity(arguments->argumentCount() + !target);
```

darinadler commented on Source/WebCore/workers/WorkerConsoleClient.cpp:
```diff
@@ -117,11 +143,86 @@ void WorkerConsoleClient::timeStamp(JSC::JSGlobalObject*, Ref<ScriptArguments>&&
+    }
+
+    Vector<JSC::Strong<JSC::Unknown>> adjustedArguments;
+    adjustedArguments.reserveInitialCapacity(arguments->argumentCount());
```
This allocates initial capacity that is 1 too small when `target` is `nullptr`, resulting in a vector that is too big when the append overflows the initial capacity. Here is one at to correct that, but maybe you can find a better way.
```suggestion
    adjustedArguments.reserveInitialCapacity(arguments->argumentCount() + !target);
```

Maybe we can share code so this tricky thing doesn't have to be written out twice?

dcrousso committed:
Web Inspector: support console screenshots in a Worker
https://bugs.webkit.org/show_bug.cgi?id=243361
<rdar://problem/98223234>

Reviewed by Darin Adler.

This will be very useful for developers that are working with `Blob` inside a `Worker` and don't have an easy way of visualizing it (e.g. sending/transferring to the webpage for rendering via the DOM).

* Source/WebCore/workers/WorkerConsoleClient.cpp:
(WebCore::canvasRenderingContext): Added.
(WebCore::WorkerConsoleClient::screenshot):
Support `ImageData`, `ImageBitmap`, `OffscreenCanvas`, all the different types of `CanvasRenderingContext`, and valid base64 `data:` URL strings.

* LayoutTests/inspector/console/console-screenshot.html:
* LayoutTests/inspector/worker/console-basic.html:
* LayoutTests/inspector/worker/console-basic-subworker.html:
* LayoutTests/inspector/worker/console-screenshot.html: Added.
* LayoutTests/inspector/worker/console-screenshot-expected.txt: Added.
* LayoutTests/inspector/worker/resources/worker-console.js:
(onmessage):

* Source/WebCore/page/FrameConsoleClient.cpp:
(WebCore::FrameConsoleClient::screenshot):

Drive-by: add some comments to clarify what each argument means.
Canonical link: https://commits.webkit.org/302778@main
